### PR TITLE
enable 'check_ldshared' in generic PythonPackage easyblock by default

### PR DESCRIPTION
**Pull request by boegel**
replacement for #1781, now to correct branch (`develop`)

### CODE REVIEW

**Comment by Micket** [comment]
I'll just copy my previous comment:
So, do we not want to make this dependent on toolchain version?
For reference:
#1455 (comment)

Alternatively, we could go through old easyconfigs where it's currently not set and set check_ldconfig=False?
We could just find all PythonPackage/PythonBundle configs for intel/iccifort/iimpi/iimpic/intelcuda that currently doesn't have `check_ldconfig` set at all, and apply `check_ldconfig = False`

**Comment by boegel** [comment]
@Micket The implementation is pretty safe, since it checks first whether the `$LDSHARED` used to build Python packages already uses the toolchain compiler.

If not (when `intel` toolchain is used together with `Python` built with `GCCcore`), only then is a different `$LDSHARED` value set.

So, I don't think we need to make this specific for certain toolchains, or go back to easyconfigs and set `check_ldshared = False` specifically.

We can do that in case having this enabled by default is problematic, but it shouldn't be (and if it is, we can still try to fix that).

We're changing this default for the upcoming EasyBuild v4.0 release, which gives us an excuse to make "breaking" changes (although nothing should actually break due to this change).

**Comment by Micket** [comment]
I know, but, but the reason why we did it "opt-in" to start with hasn't actually changed; it broke some existing old configs. These old configs relies on broken behavior of distutils.

To be specific, merging this PR would break the build for: `mpi4py-2.0.0-intel-2016b-Python-2.7.12.eb` (and several others equally old configs).

I'm not sure if we really care about these old versions (I don't use them), but, it also wouldn't be that terribly hard to fix them. Or drop them.

**Comment by boegel** [comment]
@Micket OK, just re-discovered my own fairly thorough testing of this (cfr. https://github.com/easybuilders/easybuild-easyblocks/pull/1455#issuecomment-457529348), you're right...

I'm not a big fan of hunting down easyconfigs where we should set `check_ldshared = False`, so we should come up with an additional check before we fiddle with `$LDSHARED`...

Maybe something as simple as simply implementing a bypass for any Python 2.x versions older than `2.7.15` and Python 3.x versions older than `3.7.2` is enough?

**Comment by Micket** [comment]
But we also can't put a runtime check inside `extra_options` so shall we instead do:
```python
'check_ldshared': [None, 'Check Python value of $LDSHARED, correct if needed to "$CC -shared"', CUSTOM],
```
and then, then used:
```python
if (check_ldshared is None and old_broken_versions) or check_ldshared == False:
    # Not touching LDSHARED
else:
   # Do touch LDSHARED
```
(ugh.. so ugly, perhaps a 3 value enum would be better)

**Comment by boegel** [comment]
@Micket Updated to `check_ldshared` is only set to `True` for sufficiently recent Python versions, should be good to go now?

File: easybuild/easyblocks/generic/pythonpackage.py [hunk 1]
```diff
@@ -453,6 +458,17 @@ def configure_step(self):
             except IOError:
                 raise EasyBuildError("Creating %s failed", self.sitecfgfn)
 
+        # conservatively auto-enable checking of $LDSHARED if it is not explicitely enabled or disabled
+        # only do this for sufficiently recent Python versions (>= 3.7 or Python 2.x >= 2.7.15)
+        if self.cfg.get('check_ldshared') is None:
+            pyver = det_python_version(self.python_cmd)
+            recent_py2 = pyver.startswith('2') and LooseVersion(pyver) >= LooseVersion('2.7.15')
+            if recent_py2 or LooseVersion(pyver) >= LooseVersion('3.7'):
```
Comment (Micket) [inline easybuild/easyblocks/generic/pythonpackage.py]: Do we need to go this recent?

Comment (boegel) [reply easybuild/easyblocks/generic/pythonpackage.py]: It's a conservative choice.

We're not actually building `Python` with `GCCcore` anyway (in the central repos) with older versions.

Comment (Micket) [reply easybuild/easyblocks/generic/pythonpackage.py]: Are you possibly using rpath?

**Pull request merged**